src/libratbag-data.c

```diff
@@ -207,6 +215,69 @@ init_data_hidpp20(struct ratbag *ratbag,
 	}
 }
 
+static void
+init_data_sinowealth(struct ratbag *ratbag,
+		     GKeyFile *keyfile,
+		     struct ratbag_device_data *data)
+{
+	const char *devices_group = "Driver/sinowealth/devices/";
+	const char *group = "Driver/sinowealth";
+
+	GError *error = NULL;
+
+	size_t group_count = 0;
+	_cleanup_(g_strfreevp) char **groups = g_key_file_get_groups(keyfile, &group_count);
+
+	int disable_fallback = g_key_file_get_integer(keyfile, group, "DisableFallback", &error);
+	if (!disable_fallback) {
+		data->sinowealth.fallback_device = zalloc(sizeof(struct sinowealth_device_data));
+		struct sinowealth_device_data *device = data->sinowealth.fallback_device;
+
+		device->button_count = 6;
+		device->device_name = "Generic SinoWealth Mouse";
+		device->fw_version = "0000";
+		device->led_type = SINOWEALTH_LED_TYPE_RBG;
+	}
+	g_clear_error(&error);
+
+	list_init(&data->sinowealth.supported_devices);
+
+	for (size_t i = 0; i < group_count; ++i) {
+		const char *device_group = groups[i];
+		if (!strneq(device_group, devices_group, strlen(devices_group)))
```

Comment:
fwiw, you don't need to use `strneq()` for "nice" strings, you can use `streq()` instead. `strneq()` is just for where you have at least one string that isn't NULL-terminated or where you just want to compare a prefix.

Reply:
Here I iterate over *all* groups in the device file and check if the current one is a part of `Driver/sinowealth/devices/`, not if they are completely the same, so we *do* have to use `strneq`.
Maybe there is a GLib method to get only needed groups, but I couldn't find anything. Do you know anything like that?

Reply:
ah, I misread then, apologies. I don't think there's a method for that in the GLib code (we're using .ini files a bit outside what they're originally designed for) but we *do* have `startswith()` in `shared-macro.h` which will make the code more readable for this case.

Reply:
Great, thanks!